unit/java_bytecode/generate_concrete_generic_type/generate_outer_inner.cpp

```diff
@@ -0,0 +1,367 @@
+        const symbolt &expected_symbol =
+          new_symbol_table.lookup_ref(expected_prefix);
+
+        const struct_typet class_struct = to_struct_type(expected_symbol.type);
+
+        THEN("It has field f1 of correct type")
+        {
+          const struct_union_typet::componentt &field =
+            require_type::require_component(class_struct, "f1");
+          require_type::require_pointer(
+            field.type(), symbol_typet("java::java.lang.String"));
+        }
+      }
+    }
+
+    WHEN("Its field t5 is specialised")
+    {
+      generic_utils::specialise_generic_from_component(
+        class_prefix, "t5", new_symbol_table);
+      const std::string &expected_prefix =
+        class_prefix + "$Outer$Inner$GenericInnerInner<Integer>";
+
+      THEN("The corresponding specialised class symbol is created")
+      {
+        REQUIRE(new_symbol_table.has_symbol(expected_prefix));
+        const symbolt &expected_symbol =
+          new_symbol_table.lookup_ref(expected_prefix);
+
+        const struct_typet class_struct = to_struct_type(expected_symbol.type);
+
+        THEN("It has field f1 of correct type")
+        {
+          const struct_union_typet::componentt &field =
+            require_type::require_component(class_struct, "f1");
+          require_type::require_pointer(
+            field.type(), symbol_typet("java::java.lang.Integer"));
+        }
+      }
+    }
+
+    WHEN("Its field t6 is specialised")
+    {
+      generic_utils::specialise_generic_from_component(
+        class_prefix, "t6", new_symbol_table);
+      const std::string &expected_prefix =
+        class_prefix + "$Outer$GenericInner<Integer>";
+
+      THEN("The corresponding specialised class symbol is created")
+      {
+        REQUIRE(new_symbol_table.has_symbol(expected_prefix));
+        const symbolt &expected_symbol =
+          new_symbol_table.lookup_ref(expected_prefix);
+
+        const struct_typet class_struct = to_struct_type(expected_symbol.type);
+
+        THEN("It has field f1 of correct type")
+        {
+          const struct_union_typet::componentt &field =
+            require_type::require_component(class_struct, "f1");
+          require_type::require_pointer(
+            field.type(), symbol_typet("java::generic_outer_inner$Outer$"
+                                       "GenericInner$GenericInnerInner"));
+          require_type::require_java_generic_type(
+            field.type(),
+            {{require_type::type_argument_kindt::Inst,
+              "java::java.lang.Integer"},
+             {require_type::type_argument_kindt::Inst,
+              "java::java.lang.String"}});
+        }
+      }
+    }
+
+    WHEN("Its field t7 is specialised")
+    {
+      generic_utils::specialise_generic_from_component(
+        class_prefix, "t7", new_symbol_table);
+      const std::string &expected_prefix =
+        class_prefix + "$Outer$GenericInner<Integer>$InnerInner";
```

Comment:
How come the class names aren't fully qualified - I think this might cause problems with user classes in different packages with the same names?

Reply:
Fixed.